src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2687,6 +2687,10 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
                         Error(diagnostics, ErrorCode.ERR_NoImplicitConv, syntax, distinguisher.First, distinguisher.Second);
                     }
                 }
+                else if (sourceType.SpecialType == SpecialType.System_Void)
```

Comment:
> The job of SymbolDistinguisher is to produce strings for two symbols that are different. The problem is that for this case it is asked to produce strings for two symbols that are the same.

If the problem that the two symbols are the same, then consider checking for that rather than checking for void. It would make the logic clearer and also will catch any other similar cases, should they arise in the future or under different circumstances.